### PR TITLE
refactor(transformer): use AstBuilder instance from TraverseCtx

### LINKED ISSUE #6172: Refactor transformer to use the instance of `AstBuilder` which is in `TraverseCtx` everywhere
`AstBuilder` is the object which you use to generate AST nodes. It is heavily used in the transformer.

We currently have 2 copies of `AstBuilder` in transformer:

* 1st copy is stored in `TransformCtx`
* 2nd copy is stored in `TraverseCtx`.

This is problematic for 3 reasons:

1. It blocks us from doing #6171, which will be a performance boost.
2. For the introduction of AST node IDs (#5689), `AstBuilder` will become stateful, containing a counter for creating unique IDs. So only a single copy of it can exist.
3. It's rather messy! Why do we have 2 copies anyway?

We need to go through the transformer and swap all uses of `AstBuilder` from `TransformCtx` over to use the copy in `TraverseCtx`.

For example:

https://github.com/oxc-project/oxc/blob/1cee2070506ec58a8dea86ab9ff3ca3258a98ac3/crates/oxc_transformer/src/react/jsx_source.rs#L92-L96

Should become:

```rs
let key = ctx.ast.property_key_identifier_name(SPAN, SOURCE); 
let value = self.get_source_object(line, column, ctx); 
ctx.ast.object_property_kind_object_property(SPAN, kind, key, value, None, false, false, false)
```

This will require passing `TraverseCtx` into some functions which it's currently not available in.

I have marked this as "good first issue" as it's not too technically tricky (but then again, there are always some unforeseen complications!), but is a good opportunity for anyone interested in the transformer to get to know the codebase.

Would anyone be able to help please?

### PR DESCRIPTION
**Pull request by overlookmotel**
A small modification of #6173. Closes #6172.

### CODE REVIEW

**Comment by overlookmotel** [comment]
* **#6209** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6209?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by overlookmotel** [comment]
This is basically #6173 with a few tweaks and rebased on main to solve bug fixes. I've reviewed in full, so I don't think it needs further review. So merging.

File: crates/oxc_transformer/src/lib.rs [hunk 1]
```diff
@@ -90,11 +92,12 @@ impl<'a> Transformer<'a> {
...
         program: &mut Program<'a>,
     ) -> TransformerReturn {
-        let allocator = self.ctx.ast.allocator;
+        let allocator = self.allocator;
+        let ast_builder = AstBuilder::new(allocator);
 
         let mut transformer = TransformerImpl {
             x0_typescript: TypeScript::new(self.options.typescript, &self.ctx),
-            x1_react: React::new(self.options.react, &self.ctx),
+            x1_react: React::new(self.options.react, ast_builder, &self.ctx),
```
Comment (overlookmotel) [inline crates/oxc_transformer/src/lib.rs:100]: @shulaoda This is how I solved the problem. This is pretty hacky, but it works.

We are intending to revamp the interface to `Transformer` soon, so hopefully we can come up with a more ergonomic solution then.

Comment (shulaoda) [reply crates/oxc_transformer/src/lib.rs:100]: 👍

**Pull request merged**